[digitalstrom] NPE causing multiple bindings to crash

```
2021-01-15 10:29:45.577 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NullPointerException: null
        at java.util.LinkedList$ListItr.next(LinkedList.java:893) ~[?:1.8.0_275]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.DeviceStatusManagerImpl$PollingRunnable.run(DeviceStatusManagerImpl.java:324) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_275]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_275]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
2021-01-15 10:30:01.127 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NullPointerException: null
        at java.util.LinkedList$ListItr.next(LinkedList.java:893) ~[?:1.8.0_275]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.DeviceStatusManagerImpl$PollingRunnable.run(DeviceStatusManagerImpl.java:324) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_275]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_275]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
2021-01-15 10:30:30.861 [ERROR] [rom.internal.lib.event.EventListener] - Couldn't subscribe event deviceSensorValue ... maybe timeout because system is too busy ... event will be tried to subscribe later again ...
2021-01-15 10:33:11.294 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception:
java.lang.NullPointerException: null
        at java.util.LinkedList$ListItr.next(LinkedList.java:893) ~[?:1.8.0_275]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.DeviceStatusManagerImpl$PollingRunnable.run(DeviceStatusManagerImpl.java:324) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_275]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_275]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]

```
The binding randomly throughs NPE like above. Tested with version 2.5.11.

When this exceptions occours there are other exceptions (also thrown in context of digitalstrom) that occour just a few seconds after the initial NPE. Afterwards, digitalstrom as well as some other bindings (e.g. solaredge and nibeuplink) stop working (without any further errors).
My assumption is that due to the NPE there is some http connection not closed correctly this is causing further exceptions like

```
2021-01-15 10:33:57.576 [ERROR] [rom.internal.lib.event.EventListener] - Couldn't subscribe event deviceBinaryInputEvent ... maybe timeout because system is too busy ... event will be tried to subscribe later again ...
2021-01-15 10:33:59.724 [ERROR] [verconnection.impl.HttpTransportImpl] - An IOException occurred:
javax.net.ssl.SSLException: Connection reset
        at sun.security.ssl.Alert.createSSLException(Alert.java:127) ~[?:1.8.0_275]
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:348) ~[?:1.8.0_275]
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:1.8.0_275]
        at sun.security.ssl.TransportContext.fatal(TransportContext.java:286) ~[?:1.8.0_275]
        at sun.security.ssl.SSLTransport.decode(SSLTransport.java:142) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1202) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1111) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:398) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:370) ~[?:1.8.0_275]
        at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559) ~[?:1.8.0_275]
        at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185) ~[?:1.8.0_275]
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:167) ~[?:1.8.0_275]
        at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.HttpTransportImpl.execute(HttpTransportImpl.java:256) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.HttpTransportImpl.execute(HttpTransportImpl.java:239) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.DsAPIImpl.loginApplication(DsAPIImpl.java:590) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.getNewSessionToken(ConnectionManagerImpl.java:290) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.checkConnection(ConnectionManagerImpl.java:324) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.checkConnection(ConnectionManagerImpl.java:302) [bundleFile:?]
        at org.openhab.binding.digitalstrom.internal.handler.BridgeHandler$1.run(BridgeHandler.java:594) [bundleFile:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_275]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_275]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_275]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_275]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
        Suppressed: java.net.SocketException: Broken pipe (Write failed)
                at java.net.SocketOutputStream.socketWrite0(Native Method) ~[?:1.8.0_275]
                at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111) ~[?:1.8.0_275]
                at java.net.SocketOutputStream.write(SocketOutputStream.java:155) ~[?:1.8.0_275]
                at sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:81) ~[?:1.8.0_275]
                at sun.security.ssl.TransportContext.fatal(TransportContext.java:379) ~[?:1.8.0_275]
                at sun.security.ssl.TransportContext.fatal(TransportContext.java:291) ~[?:1.8.0_275]
                at sun.security.ssl.TransportContext.fatal(TransportContext.java:286) ~[?:1.8.0_275]
                at sun.security.ssl.SSLTransport.decode(SSLTransport.java:142) ~[?:1.8.0_275]
                at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1202) ~[?:1.8.0_275]
                at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1111) ~[?:1.8.0_275]
                at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:398) ~[?:1.8.0_275]
                at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:370) ~[?:1.8.0_275]
                at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559) ~[?:1.8.0_275]
                at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185) ~[?:1.8.0_275]
                at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:167) ~[?:1.8.0_275]
                at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.HttpTransportImpl.execute(HttpTransportImpl.java:256) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.HttpTransportImpl.execute(HttpTransportImpl.java:239) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.lib.serverconnection.impl.DsAPIImpl.loginApplication(DsAPIImpl.java:590) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.getNewSessionToken(ConnectionManagerImpl.java:290) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.checkConnection(ConnectionManagerImpl.java:324) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.lib.manager.impl.ConnectionManagerImpl.checkConnection(ConnectionManagerImpl.java:302) [bundleFile:?]
                at org.openhab.binding.digitalstrom.internal.handler.BridgeHandler$1.run(BridgeHandler.java:594) [bundleFile:?]
                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_275]
                at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [?:1.8.0_275]
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_275]
                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [?:1.8.0_275]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_275]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_275]
                at java.lang.Thread.run(Thread.java:748) [?:1.8.0_275]
Caused by: java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:210) ~[?:1.8.0_275]
        at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:476) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketInputRecord.readHeader(SSLSocketInputRecord.java:470) ~[?:1.8.0_275]
        at sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:160) ~[?:1.8.0_275]
        at sun.security.ssl.SSLTransport.decode(SSLTransport.java:110) ~[?:1.8.0_275]
        ... 21 more
```
This might break the functionality of the shared http-client which is also used by solaredge and nibeuplink.

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMHandlerFactory.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMHandlerFactory.java
@@ -16,6 +16,7 @@ import static org.openhab.binding.digitalstrom.internal.DigitalSTROMBindingConst
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.commons.lang.StringUtils;
 import org.openhab.binding.digitalstrom.internal.discovery.DiscoveryServiceManager;
@@ -50,7 +51,7 @@ import org.slf4j.LoggerFactory;
 public class DigitalSTROMHandlerFactory extends BaseThingHandlerFactory {
 
     private final Logger logger = LoggerFactory.getLogger(DigitalSTROMHandlerFactory.class);
-    private final Map<String, DiscoveryServiceManager> discoveryServiceManagers = new HashMap<>();
+    private final Map<String, DiscoveryServiceManager> discoveryServiceManagers = new ConcurrentHashMap<>();
 
     private Map<ThingUID, BridgeHandler> bridgeHandlers;
 
@@ -256,10 +257,9 @@ public class DigitalSTROMHandlerFactory extends BaseThingHandlerFactory {
     protected synchronized void removeHandler(ThingHandler thingHandler) {
         if (thingHandler instanceof BridgeHandler) {
             String uid = thingHandler.getThing().getUID().getAsString();
-            DiscoveryServiceManager discoveryServiceManager = discoveryServiceManagers.get(uid);
+            DiscoveryServiceManager discoveryServiceManager = discoveryServiceManagers.remove(uid);
             if (discoveryServiceManager != null) {
                 discoveryServiceManager.unregisterDiscoveryServices(bundleContext);
-                discoveryServiceManagers.remove(uid);
             }
         }
     }

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMHandlerFactory.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/DigitalSTROMHandlerFactory.java
@@ -256,8 +256,9 @@ public class DigitalSTROMHandlerFactory extends BaseThingHandlerFactory {
     protected synchronized void removeHandler(ThingHandler thingHandler) {
         if (thingHandler instanceof BridgeHandler) {
             String uid = thingHandler.getThing().getUID().getAsString();
-            if (discoveryServiceManagers.get(uid) != null) {
-                discoveryServiceManagers.get(uid).unregisterDiscoveryServices(bundleContext);
+            DiscoveryServiceManager discoveryServiceManager = discoveryServiceManagers.get(uid);
+            if (discoveryServiceManager != null) {
+                discoveryServiceManager.unregisterDiscoveryServices(bundleContext);
                 discoveryServiceManagers.remove(uid);
             }
         }

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
@@ -19,6 +19,7 @@ import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
@@ -113,7 +114,7 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
     private SceneReadingJobExecutor sceneJobExecutor;
     private EventListener eventListener;
 
-    private final List<TrashDevice> trashDevices = new LinkedList<>();
+    private final List<TrashDevice> trashDevices = new CopyOnWriteArrayList<>();
 
     private long lastBinCheck = 0;
     private ManagerStates state = ManagerStates.STOPPED;
@@ -310,7 +311,6 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
                             }
                         }
                     }
-
                 } else {
                     logger.debug("Found new device!");
                     if (trashDevices.isEmpty()) {
@@ -320,26 +320,19 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
                                 currentDevice.getDSID());
                     } else {
                         logger.debug("Search device in trashDevices.");
-                        TrashDevice foundTrashDevice = null;
-                        for (TrashDevice trashDevice : trashDevices) {
-                            if (trashDevice != null) {
-                                if (trashDevice.getDevice().equals(currentDevice)) {
-                                    foundTrashDevice = trashDevice;
+                        trashDevices.stream().filter(d -> d.getDevice().equals(currentDevice)).findAny()
+                                .ifPresentOrElse(trashDevice -> {
                                     logger.debug(
                                             "Found device in trashDevices, add TrashDevice with dSID {} to the StructureManager!",
                                             currentDeviceDSID);
-                                }
-                            }
-                        }
-                        if (foundTrashDevice != null) {
-                            trashDevices.remove(foundTrashDevice);
-                            strucMan.addDeviceToStructure(foundTrashDevice.getDevice());
-                        } else {
-                            strucMan.addDeviceToStructure(currentDevice);
-                            logger.debug(
-                                    "Can't find device in trashDevices, add Device with dSID: {} to the StructureManager!",
-                                    currentDeviceDSID);
-                        }
+                                    trashDevices.remove(trashDevice);
+                                    strucMan.addDeviceToStructure(trashDevice.getDevice());
+                                }, () -> {
+                                    strucMan.addDeviceToStructure(currentDevice);
+                                    logger.debug(
+                                            "Can't find device in trashDevices, add Device with dSID: {} to the StructureManager!",
+                                            currentDeviceDSID);
+                                });
                     }
                     if (deviceDiscovery != null) {
                         // only informs discovery, if the device is a output or a sensor device
@@ -387,18 +380,16 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
                             DeviceStatusListener.DEVICE_DISCOVERY, device.getDSID().getValue());
                 } else {
                     logger.debug(
-                            "The device-Discovery is not registrated, can't inform device discovery about removed device.");
+                            "The device-Discovery is not registered, can't inform device discovery about removed device.");
                 }
             }
 
             if (!trashDevices.isEmpty() && (lastBinCheck + config.getBinCheckTime() < System.currentTimeMillis())) {
-                for (TrashDevice trashDevice : trashDevices) {
-                    if (trashDevice.isTimeToDelete(Calendar.getInstance().get(Calendar.DAY_OF_YEAR))) {
-                        logger.debug("Found trashDevice that have to delete!");
-                        trashDevices.remove(trashDevice);
-                        logger.debug("Delete trashDevice: {}", trashDevice.getDevice().getDSID().getValue());
-                    }
-                }
+                trashDevices.stream().filter(d -> d.isTimeToDelete(Calendar.getInstance().get(Calendar.DAY_OF_YEAR)))
+                        .forEach(trashDevice -> {
+                            trashDevices.remove(trashDevice);
+                            logger.debug("Deleted trashDevice: {}", trashDevice.getDevice().getDSID().getValue());
+                        });
                 lastBinCheck = System.currentTimeMillis();
             }
         }

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
@@ -320,19 +320,24 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
                                 currentDevice.getDSID());
                     } else {
                         logger.debug("Search device in trashDevices.");
-                        trashDevices.stream().filter(d -> d.getDevice().equals(currentDevice)).findAny()
-                                .ifPresentOrElse(trashDevice -> {
-                                    logger.debug(
-                                            "Found device in trashDevices, add TrashDevice with dSID {} to the StructureManager!",
-                                            currentDeviceDSID);
-                                    trashDevices.remove(trashDevice);
-                                    strucMan.addDeviceToStructure(trashDevice.getDevice());
-                                }, () -> {
-                                    strucMan.addDeviceToStructure(currentDevice);
-                                    logger.debug(
-                                            "Can't find device in trashDevices, add Device with dSID: {} to the StructureManager!",
-                                            currentDeviceDSID);
-                                });
+                        boolean found = trashDevices.removeIf(trashDevice -> {
+                            if (trashDevice.getDevice().equals(currentDevice)) {
+                                logger.debug(
+                                        "Found device in trashDevices, add TrashDevice with dSID {} to the StructureManager!",
+                                        currentDeviceDSID);
+                                strucMan.addDeviceToStructure(trashDevice.getDevice());
+                                return true;
+                            } else {
+                                return false;
+                            }
+                        });
+                        if (!found) {
+                            strucMan.addDeviceToStructure(currentDevice);
+                            logger.debug(
+                                    "Can't find device in trashDevices, add Device with dSID: {} to the StructureManager!",
+                                    currentDeviceDSID);
+                        }
+                        ;
                     }
                     if (deviceDiscovery != null) {
                         // only informs discovery, if the device is a output or a sensor device
@@ -385,11 +390,14 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
             }
 
             if (!trashDevices.isEmpty() && (lastBinCheck + config.getBinCheckTime() < System.currentTimeMillis())) {
-                trashDevices.stream().filter(d -> d.isTimeToDelete(Calendar.getInstance().get(Calendar.DAY_OF_YEAR)))
-                        .forEach(trashDevice -> {
-                            trashDevices.remove(trashDevice);
-                            logger.debug("Deleted trashDevice: {}", trashDevice.getDevice().getDSID().getValue());
-                        });
+                trashDevices.removeIf(trashDevice -> {
+                    if (trashDevice.isTimeToDelete(Calendar.getInstance().get(Calendar.DAY_OF_YEAR))) {
+                        logger.debug("Deleted trashDevice: {}", trashDevice.getDevice().getDSID().getValue());
+                        return true;
+                    } else {
+                        return false;
+                    }
+                });
                 lastBinCheck = System.currentTimeMillis();
             }
         }

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/DeviceStatusManagerImpl.java
@@ -337,7 +337,6 @@ public class DeviceStatusManagerImpl implements DeviceStatusManager {
                                     "Can't find device in trashDevices, add Device with dSID: {} to the StructureManager!",
                                     currentDeviceDSID);
                         }
-                        ;
                     }
                     if (deviceDiscovery != null) {
                         // only informs discovery, if the device is a output or a sensor device

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/SceneManagerImpl.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/SceneManagerImpl.java
@@ -286,15 +286,16 @@ public class SceneManagerImpl implements SceneManager {
             }
         } else {
             InternalScene oldScene = this.internalSceneMap.get(intScene.getID());
-            String oldSceneName = this.internalSceneMap.get(intScene.getID()).getSceneName();
-            String newSceneName = intScene.getSceneName();
-            if (oldScene != null
-                    && (oldSceneName.contains("Zone:") && oldSceneName.contains("Group:")
-                            && oldSceneName.contains("Scene:"))
-                    && !(newSceneName.contains("Zone:") && newSceneName.contains("Group:")
-                            && newSceneName.contains("Scene:"))) {
-                oldScene.setSceneName(newSceneName);
-                this.discovery.sceneDiscoverd(oldScene);
+            if (oldScene != null) {
+                String oldSceneName = oldScene.getSceneName();
+                String newSceneName = intScene.getSceneName();
+                if ((oldSceneName.contains("Zone:") && oldSceneName.contains("Group:")
+                        && oldSceneName.contains("Scene:"))
+                        && !(newSceneName.contains("Zone:") && newSceneName.contains("Group:")
+                                && newSceneName.contains("Scene:"))) {
+                    oldScene.setSceneName(newSceneName);
+                    this.discovery.sceneDiscoverd(oldScene);
+                }
             }
         }
     }

--- a/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/SceneManagerImpl.java
+++ b/bundles/org.openhab.binding.digitalstrom/src/main/java/org/openhab/binding/digitalstrom/internal/lib/manager/impl/SceneManagerImpl.java
@@ -288,7 +288,9 @@ public class SceneManagerImpl implements SceneManager {
             InternalScene oldScene = this.internalSceneMap.get(intScene.getID());
             String oldSceneName = this.internalSceneMap.get(intScene.getID()).getSceneName();
             String newSceneName = intScene.getSceneName();
-            if ((oldSceneName.contains("Zone:") && oldSceneName.contains("Group:") && oldSceneName.contains("Scene:"))
+            if (oldScene != null
+                    && (oldSceneName.contains("Zone:") && oldSceneName.contains("Group:")
+                            && oldSceneName.contains("Scene:"))
                     && !(newSceneName.contains("Zone:") && newSceneName.contains("Group:")
                             && newSceneName.contains("Scene:"))) {
                 oldScene.setSceneName(newSceneName);

